### ISSUE TITLE
Broken fonts in latest iTerm2 version

### ISSUE THREAD

**Issue by jfeaver** [powerline/fonts#44]
I've tried using Inconsolata and SourceCodePro.  I've also opened an issue in iTerm2 (https://code.google.com/p/iterm2/issues/detail?id=2518&thanks=2518&ts=1372092283).

What steps will reproduce the problem?
1. Download iTerm2 previous stable release (early 2013 version??)
2. Download and install powerline, configured for vim/tmux: https://github.com/Lokaltog/powerline
2a. Notice that powerline does not look visually pleasing (it has missing character marks all over).
3. Download and setup iTerm2 to use a powerline font: https://github.com/Lokaltog/powerline-fonts
3b. Notice that the new font makes the powerline in vim look visually pleasing (as in vim-powerline.png)
4. Upgrade iTerm2 to latest stable release 1.0.0.20130622
5. Open vim or tmux in iTerm2 and notice broken font

What is the expected output? What do you see instead?

I expect that iTerm2 will use the patched font while running powerline to display a visual interface as in the attached screenshot (vim-powerline.png).

Instead, I am seeing unknown character marks as in the attached screenshot (Screen Shot ... .png)

What version of the product are you using? On what operating system?
iTerm2 1.0.0.20130622
Mac OS X 10.8.4 Build 12E55

Please provide any additional information below.
![vim-powerline](https://f.cloud.github.com/assets/1091338/696715/85133e70-dced-11e2-9806-fc79ede3e880.png)
![screen shot 2013-06-24 at 10 33 01 am](https://f.cloud.github.com/assets/1091338/696717/8dd27350-dced-11e2-8615-5d8fc23a2346.png)

**Comment by apjanke** [comment]
Hi folks. Is this issue outstanding, or did you determine it was the code point change and the current version of the fonts fixes it?

**Comment by iToto** [comment]
Resolved here - although I forgot what I did to resolve it :fearful:

**Comment by iwrightcode** [comment]
Hey all. I'm with apjanke. I am having that issue still as well. Any answers?

**Comment by apjanke** [comment]
Actually, it's working fine for me now: we (oh-my-zsh workers) determined that it was the "agnoster" theme in `oh-my-zsh` which was using outdated code points, and the current version of these patched fonts is correct, with the current Powerline private code points. I got my problems fixed by getting oh-my-zsh updated to use the current Powerline code points in its theme definitions, and updating my local oh-my-zsh installations.

I think this issue can be closed.

@iwrightcode: What specific issues are you having now? If I'm correct in my understanding of this, they can probably be cleared up by a) downloading the latest version of these powerline fonts, b) making sure your theme and terminal definitions are up to date with the current code Point values, and c) setting your "non-ASCII" font in iTerm to one of the Powerline fonts.

**Comment by kamontat** [comment]
I try this fonts and this bug become again

# Version
MacOS: 10.12.4
iTerm2: 3.0.15

![screen shot 2560-05-03 at 1 07 19 am](https://cloud.githubusercontent.com/assets/14089557/25632261/2ab375ba-2f9d-11e7-831f-a2865e1cdf9c.png)

**Comment by bperucchi** [comment]
Sorry, but did try many resolution in this post but still not get icons correctly in prompt line.

I'm using "Meslo LG S Bold for Powerline" font on Mac Sierra 

<img width="1366" alt="screen shot 2017-05-08 at 02 07 00" src="https://cloud.githubusercontent.com/assets/367967/25791035/01137714-3394-11e7-9c53-c1798f6a4843.png">


<img width="1348" alt="screen shot 2017-05-08 at 03 30 17" src="https://cloud.githubusercontent.com/assets/367967/25792751/eb004a14-339e-11e7-8b25-0f83bd993a76.png">

**Comment by fusionx1** [comment]
hi @bperucchi Im experiencing the same issue. Im also looking for an answer, lets share some notes.   So far I already done adding powerline fonts but its still not working.

**Comment by romandoma** [comment]
Hi @bperucchi @fusionx1 

I had the same problem, with the terminal exactly the same as @bperucchi. I solved it with the following configuration:

![image](https://cloud.githubusercontent.com/assets/4895466/25925757/4e7f0aa4-35b0-11e7-9b30-e8b8fc392dbf.png)

**Comment by bperucchi** [comment]
@fusionx1 @romandoma @kamontat 

Thanks for your reply but I'm still get the same problem in the icons

<img width="996" alt="screen shot 2017-05-11 at 02 14 42" src="https://cloud.githubusercontent.com/assets/367967/25933805/4db645b6-35f0-11e7-8d89-c41d94264b71.png">

**Comment by tmikulin** [comment]
@bperucchi, I know its a long shot, but have you restarted your Iterm2? If you installed everything correctly, that could be the problem. Before restarting Iterm2 just open a second tab and see if the icons are showing.

**Comment by bperucchi** [comment]
@tmikulin already did that and still got the icons problem. :/

**Comment by Helcaraxan** [comment]
Same story on my side. Having issue in iTerm2 alone and they are intermittent.

The broken down output:
![screen shot 2017-05-17 at 09 16 08](https://cloud.githubusercontent.com/assets/896592/26144648/910479dc-3ae1-11e7-9485-507c14b6d5bb.png)

Sometimes rebooting iTerm2 fixes the issue, sometimes rebooting my mac fixes it and sometimes both fail to do anything. I have triple checked all setting as well as the patched fonts and nothing suspicious there: everything looks as it's supposed to be.

Even when the iTerm2 fonts are not working it still works in Vim inside that same window.

**Comment by bperucchi** [comment]
I solved my problem! I I just open mac "Font Book" and repair the fonts.

> I actually resolved this for myself after several frustrating hours of toying with Powerline and such. Essentially I had to go into Font Book and resolve "issues" with the powerline fonts I had installed (perhaps because I tried installing/reinstalling them several times? Either way - first install didn't work either). Inside Font Book I chose to allow it to repair the issue automatically for each font - the special characters now display correctly.

https://github.com/robbyrussell/oh-my-zsh/issues/2869

@fusionx1 @romandoma @kamontat

**Comment by Helcaraxan** [comment]
Thanks for that point @brenoperucchi. I'll try it out the next time my fonts break. Currently after 1 or 2 more reboots they seem to have come back to their senses automatically.

**Comment by b-gran** [comment]
**TLDR; `powerline-daemon --replace` fixes the problem for me**

I've been having this problem intermittently for the past few months. Some observations:
1. powerline works fine in vim within iTerm2, but bash still has the font problem
2. `powerline-render shell left` within bash has the correct behavior (i.e. no font issues)
3. `powerline-daemon --replace` fixes the font issue for some amount of time, until eventually the font issues crop up again

**Comment by Helcaraxan** [comment]
So I tried the "Fix fonts" proposed by @brenoperucchi but it did not work for me. I do see that I am in exactly the same case as what you describe @b-gran w.r.t vim and bash in iTerm2. And the `powerline-daemon --replace` does fix it! This points to something getting corrupt within the daemon but not in the OSX fonts itself.

Will need to take a closer look into Powerline's source code to see what is going wrong. Potentially a cache that gets corrupted.

**Comment by ZyX-I** [comment]
@Helcaraxan The only thing which may *look* like a font issue with daemon is daemon launched in non-UTF-8 environment and thus assuming that Unicode is not supported. This has nothing to do with fonts actually, just it should choose `ascii` theme under given circumstances.

**Comment by Helcaraxan** [comment]
@ZyX-I but would that explain why the issue starts appearing randomly while running even if it works correctly initially? What you describe looks like something that should be binary on restart: either it always works or it never does.

**Comment by ZyX-I** [comment]
@Helcaraxan No idea actually. It should work just like this: always or never. It may replace powerline triangles with question marks if you force theme without fixing the encoding (there are ways to do this via environment variables), but it will not recheck locale.

**Comment by ZyX-I** [comment]
@Helcaraxan But daemon does not do anything with fonts, it receives text data (current working directory, arguments to the client and the whole client environment), processes it to generate a prompt and pipes generated prompt back to the client. Your screenshot is exactly what daemon should pipe back should it be non-unicode.

About note: there is `locale`, I do not know what actually “iTerm settings” do, but what is needed to do is setting environment variables. On linux it is normally something like `LANG=en_US.UTF-8` and it is set in `/etc/profile`. I saw some OS X users going with LC_ALL instead for some reason. The `locale` command prints what locale is.

**Comment by ZyX-I** [comment]
Based on some other terminal’s settings all setting `UTF-8` in their settings does is making terminal emulator interpret input to the terminal as unicode. And *not* setting locale. Actually do not expect any setting that is a mere “set encoding to something” to set locale: locale is *encoding* **plus** *language* (latter defines collation rules, decimal separator, etc). If you see a setting which is missing “language” part then know that it has nothing to do with locale.

**Comment by Helcaraxan** [comment]
So the only difference here is that in my case (OS X) I have the following:
```sh
 -> locale
LANG="en_GB.UTF-8"
LC_COLLATE="en_GB.UTF-8"
LC_CTYPE="en_GB.UTF-8"
LC_MESSAGES="en_GB.UTF-8"
LC_MONETARY="en_GB.UTF-8"
LC_NUMERIC="en_GB.UTF-8"
LC_TIME="en_GB.UTF-8"
LC_ALL=
```

**Comment by ZyX-I** [comment]
@Helcaraxan You need to check this at the time daemon is being run. E.g. replace `powerline-daemon` executable with a script which runs `pstree -h` (to check who runs the daemon) and `locale` (to actually check locale), saving theirs output to some known location and then runs actual `powerline-daemon` executable.

**Comment by Helcaraxan** [comment]
I understand yeah and will provide the output of that action when possible.

However I again do not see why this would cause the issue that this topic describes. Either the `locale` is set correctly on start and it works or it is not set correctly and fails immediately. I have never heard of the environment variables of a daemonized process changing while running unless the process itself does that or if you start attaching a debugger or other gnarly approaches like that.

Don't get me wrong: I am not trying to contradict you _per se_, just that I am really wondering what the underlying issue is as, IMHO, your suggestions would not account for the exact behaviour that is being observed by @b-gran, me and others on this topic, _i.e_ that the prompt displays correctly initially and then gets corrupted after some (potentially long, multiple day) period.

**Comment by ZyX-I** [comment]
@Helcaraxan I would also add `date` output there then (and both before and after calling `powerline-daemon`). It is actually possible that daemon was killed and restarted, though this is also strange as with all documented ways of using daemon you need to start a new application (shell or tmux *server* instance) to get daemon restarted while without daemon `powerline` will use the shell locale.

**Comment by Helcaraxan** [comment]
Ok. Will do. It might take a while before I will be able to post the result as I'll need to wait until the display gets corrupted again but I won't miss on following up.

In any case thank you very much for the attention that you are giving to the issue!

**Comment by Helcaraxan** [comment]
@ZyX-I the issue just reappeared which gave me [the requested log output](https://github.com/powerline/fonts/files/1185876/powerline-boot-log.txt). Unfortunately it does not look like it gives up any extra information.

It happened after I opened up a new tab in iTerm2 (screenshot of the resulting aspect below). Running `powerline-daemon --replace` solved it immediately (also see screenshot).

<img width="1434" alt="powerline-font-issue" src="https://user-images.githubusercontent.com/896592/28755708-8949e9fa-7558-11e7-8427-299ef60cff1f.png">

**Comment by ZyX-I** [comment]
@Helcaraxan You have

```
Starting powerline-daemon with args:  -q
Sun Jul 30 18:44:12 BST 2017
Locale
LANG=
LC_COLLATE="C"
LC_CTYPE="C"
LC_MESSAGES="C"
LC_MONETARY="C"
LC_NUMERIC="C"
LC_TIME="C"
LC_ALL=
End of log
```

right above “the issue appeared with the launch below”, in the same minute, and a bunch of the same runs (with C locale as well) throughtout the log. Do not forget that `powerline-daemon -i` will *not* start the daemon if it is already running.

**Comment by Helcaraxan** [comment]
All these runs are implicit. I have not triggered any of them apart from the last one with `--replace`.

**Comment by ZyX-I** [comment]
@Helcaraxan I have not said you did. I just meant that “the issue appeared with the launch below” is most likely pointing to the wrong launch because it is *previous* one which actually started the daemon, and it did start it with the wrong locale.

By the way, your script has an error: `'$$'` is literal `$$` and I guess you wanted a PID.

**Comment by ZyX-I** [comment]
@Helcaraxan In any case, normally you must have exactly zero runs with C locale.

**Comment by ZyX-I** [comment]
It appears that current discussion has absolutely nothing to do with the original post.

**Comment by ZyX-I** [comment]
Since with C locale it is not a powerline *issue* better move to the mailing list.
